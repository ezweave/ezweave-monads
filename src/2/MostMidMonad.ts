interface IMonad<A> {
  bind<B>(callback: (value?: A) => Monad<B>): IMonad<B>;
}

abstract class Monad<A> implements IMonad<A> {
  abstract bind<B>(fn: (value?: A) => IMonad<B>): IMonad<B>;

  public static some<A>(value: A) {
    return new MostMidMonad<A>(value);
  }

  public static none() {
    return new MostMidMonad<undefined>();
  }
}

export class MostMidMonad<A> extends Monad<A> {
  myValue?: A;

  constructor(t?: A) {
    super();
    if (t) {
      this.myValue = t;
    }
  }

  public bind<B>(callback: (value?: A) => IMonad<B>) {
    return callback(this.myValue);
  }
}
