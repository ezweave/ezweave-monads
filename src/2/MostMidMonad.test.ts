import { toUpper } from 'lodash';

import { MostMidMonad } from './MostMidMonad';

describe(MostMidMonad, () => {
  it('can use some to create a new Monad statically', () => {
    const one = MostMidMonad.some('foo');

    const two = one.bind(
      (value) => new MostMidMonad(`${toUpper(value)}_SO_MID`),
    );

    expect(two).toMatchSnapshot();
  });
  it('can use none to create a new Monad statically', () => {
    const one = MostMidMonad.none();

    const two = one.bind(
      (value) => new MostMidMonad(`${toUpper(value)}_SO_MID`),
    );

    expect(two).toMatchSnapshot();
  });
  it('can bind a bunch of functions', () => {
    const one = MostMidMonad.some('foo');

    const two = one
      .bind((value) => new MostMidMonad(toUpper(value)))
      .bind((value) => new MostMidMonad(`${value}_SO_MID`));

    expect(two).toMatchSnapshot();
  });
});
