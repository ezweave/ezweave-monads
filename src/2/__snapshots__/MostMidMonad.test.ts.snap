// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MostMidMonad can bind a bunch of functions 1`] = `
MostMidMonad {
  "myValue": "FOO_SO_MID",
}
`;

exports[`MostMidMonad can use none to create a new Monad statically 1`] = `
MostMidMonad {
  "myValue": "_SO_MID",
}
`;

exports[`MostMidMonad can use some to create a new Monad statically 1`] = `
MostMidMonad {
  "myValue": "FOO_SO_MID",
}
`;
